kustomize `buildMetadata` documentation (#236)
kubernetes-sigs/cli-experimental: site/content/en/references/kustomize/kustomization/buildmetadata/_index.md

---
title: "buildMetadata"
linkTitle: "buildMetadata"
type: docs
weight: 18
description: >
    Specify options for including information about the build in annotations or labels. 
---

The `buildMetadata` field is a list of strings. The strings can be one of three builtin
options that add some metadata to each resource about how the resource was built. 

These options are:

- `managedByLabel`
- `originAnnotations`
- `transformerAnnotations`

It is possible to set one or all of these options in the kustomization file:

> It is possible for to set both `transformerAnnotations` and `originAnnotations`:

Reviewer: Remove? Same as above but missing managedByLabel.

Author: Ah my bad, copy paste mistake. Thanks for the catch!

```yaml
buildMetadata: [managedByLabel, originAnnotations, transformerAnnotations]
```

### Managed By Label
To mark the resource as having been managed by kustomize, you can specify the `managedByLabel`
option in the `buildMetadata` field of the kustomization:

```yaml
buildMetadata: [managedByLabel]
```

> This will add a label `app.kubernetes.io/managed-by` to each resource with the version of kustomize 

Reviewer: ```suggestion
This will add the label `app.kubernetes.io/managed-by` to each resource with the version of kustomize 
```

that has built it. For example, given the following input:

kustomization.yaml
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- service.yaml
buildMetadata: [managedByLabel]
```

service.yaml
```yaml
apiVersion: v1
kind: Service
metadata:
  name: myService
spec:
  ports:
  - port: 7002
```

`kustomize build` will produce a resource with an output like the following:

```yaml
apiVersion: v1
kind: Service
metadata:
  labels:

>     app.kubernetes.io/managed-by: kustomize-v444.333.222

Reviewer: nit: use a non-dev version for the example

  name: myService
spec:
  ports:
  - port: 7002
```

### Origin Annotation

> To retain the information about the origin of a resource, you can specify the `originAnnotations` 

Reviewer: ```suggestion
To annotate resources with information about their origin, you can specify the `originAnnotations` 
```

option in the `buildMetadata` field of the kustomization: 

```yaml
buildMetadata: [originAnnotations]
```

> When this option is set, generated resources should receive an annotation with key `config.kubernetes.io/origin`,

Reviewer: ```suggestion
When this option is set, generated resources will receive an annotation with key `config.kubernetes.io/origin`,
```

containing data about the generator that produced it. If the resource is from the `resources` field, this annotation
contains data about the file it originated from. 

> The possible fields of these annotations are:

Reviewer: Can we break this out into an example and a separate field list? It's a bit hard to read since the comments aren't syntax highlighted (they're just part of the string).

![image](https://user-images.githubusercontent.com/4789493/152072738-62f3fb9a-7fde-41f7-9e2b-a7fdfb2b3785.png)

```yaml
config.kubernetes.io/origin: |
  path: path.yaml # The path to the resource file itself
  ref: v0.0.0 # If from a remote file or generator, the ref of the repo URL
  repo: http://github.com/examplerepo # If from a remote file or generator, the repo source
  configuredIn: path/to/generatorconfig # If a generated resource, the path to the generator config
  configuredBy: # If a generated resource, the ObjectReference of the generator config
    kind: Generator
    apiVersion: builtin 
    name: foo
    namespace: default
```

All local file paths are relative to the top-level kustomization, i.e. the kustomization file in the directory upon 
which `kustomize build` was invoked. For example, if someone were to run `kustomize build foo`, all file paths 
in the annotation output would be relative to`foo/kustomization.yaml`. 

All remote file paths are relative to the root of the remote repository. 

Any fields that are not applicable would be omitted from the final output. If a generator is invoked via 

> a field in the kustomization file, `configuredIn` would point to the kustomization file itself. 

Reviewer: If you split out the field list, I suggest also moving this to the field explanation itself.

```yaml
config.kubernetes.io/origin: |
  configuredIn: generator.yaml
  configuredBy: 
    kind: MyGenerator
    apiVersion: v1
    name: generator
```

##### Remote builtin generator
We have a top-level kustomization such as the following:

```yaml
resources:
- github.com/examplerepo/?ref=v1.0.6

buildMetadata: [originAnnotations]
```

which uses `github.com/examplerepo/?ref=v1.0.6` as a remote base. This remote base has the following kustomization
defined in `github.com/examplerepo/kustomization.yaml`:

```yaml
configMapGenerator:
- name: my-java-server-env-vars
  literals:
  - JAVA_HOME=/opt/java/jdk
  - JAVA_TOOL_OPTIONS=-agentlib:hprof
```

Running `kustomize build` on the top-level kustomization would produce the following output:

```yaml
apiVersion: v1
data:
  JAVA_HOME: /opt/java/jdk
  JAVA_TOOL_OPTIONS: -agentlib:hprof
kind: ConfigMap
metadata:
  name: my-java-server-env-vars-44k658k8gk
  annotations:
    config.kubernetes.io/origin: |
      ref: v1.0.6 
      repo: github.com/examplerepo 
      configuredIn: kustomization.yaml
      configuredBy: 
        kind: ConfigMapGenerator
        apiVersion: builtin 
```

### Transformer Annotations [Alpha]

> To retain the information about what transformers have acted on each resource, you can add the

Reviewer: ```suggestion
To annotate resources with information about the transformers that have acted on them, you can add the
```

`transformerAnnotations` option to the `buildMetadata` field of the kustomization:

```yaml
buildMetadata: [transformerAnnotations]
```

When the `transformerAnnotations` option is set, kustomize will add annotations with information about what transformers 
have acted on each resource. Transformers can be invoked either through various fields in the kustomization file 
(e.g. the `replacements` field will invoke the ReplacementTransformer), or through the `transformers` field.

The annotation key for transformer annotations will be `config.kubernetes.io/transformations`, which will contain a list of
transformer data:

```yaml
config.kubernetes.io/transformations: |

> - ref: v0.0.0 # If from a remote transformer, the ref of the repo URL

Reviewer: I think this should be indented one, plus I have the same suggestion about separating the example from the field list because of the comment syntax highlighting issue.

  repo: http://github.com/examplerepo # If from a remote transformer, the repo source
  configuredIn: path/to/transformerconfig # The path to the transformer config
  configuredBy: # The ObjectReference of the transformer config
    kind: Transformer
    apiVersion: builtin 
    name: foo
    namespace: default
```

While this field is in alpha, it will receive the `alpha` prefix, so you will see the annotation key 

> `alpha.config.kubernetes.io/transformations` instead. 

Reviewer: I think I asked for this, but did we decide that we are not guaranteeing that the annotation content will be stable during alpha? If so, we should mention that.

The possible fields for each item in the list is identical to the possible fields in `config.kubernetes.io/origin`, except that
the transformer annotation does not have a `path` field for the path to the resource file itself. 

All local file paths are relative to the top-level kustomization, i.e. the kustomization file in the directory upon 
which `kustomize build` was invoked. For example, if someone were to run `kustomize build foo`, all file paths 
in the annotation output would be relative to`foo/kustomization.yaml`. 

All remote file paths are relative to the root of the remote repository. 

Any fields that are not applicable would be omitted from the final output. If a transformer is invoked via 

> a field in the kustomization file, `configuredIn` would point to the kustomization file itself. 

Reviewer: Same comment re: moving this to be with the other configuredIn info
